contrib/hyperscan/matching/input_matchers/source/matcher.h

```diff
@@ -14,26 +13,42 @@ namespace InputMatchers {
 namespace Hyperscan {
 
 struct ScratchThreadLocal : public ThreadLocal::ThreadLocalObject {
-  explicit ScratchThreadLocal(const hs_database_t* database);
-  ~ScratchThreadLocal() override { hs_free_scratch(scratch_); }
+  ScratchThreadLocal(const hs_database_t* database, const hs_database_t* som_database);
+  ~ScratchThreadLocal() override;
+
   hs_scratch_t* scratch_{};
 };
 
-class Matcher : public Envoy::Matcher::InputMatcher {
+struct Bound {
+  Bound(uint64_t start, uint64_t end);
+
+  bool operator<(const Bound& other) const;
+
+  uint64_t start_;
+  uint64_t end_;
+};
+
+class Matcher : public Envoy::Regex::CompiledMatcher, public Envoy::Matcher::InputMatcher {
 public:
-  explicit Matcher(
-      const envoy::extensions::matching::input_matchers::hyperscan::v3alpha::Hyperscan& config,
-      ThreadLocal::SlotAllocator& tls);
-  ~Matcher() override { hs_free_database(database_); }
+  Matcher(const std::vector<const char*>& expressions, const std::vector<unsigned int>& flags,
+          const std::vector<unsigned int>& ids, ThreadLocal::SlotAllocator& tls,
+          bool report_start_of_matching);
+  ~Matcher() override;
+
+  // Envoy::Regex::CompiledMatcher
+  bool match(absl::string_view value) const override;
+  std::string replaceAll(absl::string_view value, absl::string_view substitution) const override;
 
   // Envoy::Matcher::InputMatcher
   bool match(absl::optional<absl::string_view> input) override;
 
 private:
-  std::vector<unsigned int> flags_{};
-  std::vector<unsigned int> ids_{};
   hs_database_t* database_{};
+  hs_database_t* som_database_{};
```

Comment:
ideally we should spell out whatever som stands for everywhere by renaming "som" as a variable name segment. but at a minimum maybe expand that acronym here?

Reply:
Done.